include/libtorrent/aux_/rtc_signaling.hpp

```diff
@@ -0,0 +1,173 @@
+#include <memory>
+#include <optional>
+#include <queue>
+#include <vector>
+#include <chrono>
+
+namespace rtc {
+	class PeerConnection;
+	class DataChannel;
+}
+
+namespace libtorrent {
+
+class alert_manager;
+class torrent;
+
+namespace aux {
+
+struct rtc_stream_init;
+
+const int RTC_OFFER_ID_LEN = 16;
+
+struct rtc_offer_id : std::vector<char> {
+	rtc_offer_id() : std::vector<char>(RTC_OFFER_ID_LEN, '\0') {}
+	explicit rtc_offer_id(span<char const> s) : std::vector<char>(s.begin(), s.end()) {}
+};
+
+struct rtc_offer_id_hash
+{
+	std::size_t operator()(rtc_offer_id const& id) const
+	{
+        return boost::hash<std::vector<char>>{}(id);
+    }
+};
+
+struct rtc_answer
+{
+	rtc_offer_id offer_id;
+	peer_id pid;
+	std::string sdp;
+};
+
+struct rtc_offer
+{
+	rtc_offer_id id;
+	peer_id pid;
+	std::string sdp;
+	std::function<void(peer_id const& pid, rtc_answer const& answer)> answer_callback;
+};
+
+// This class handles client signaling for WebRTC DataChannels
+class TORRENT_EXTRA_EXPORT rtc_signaling final : public std::enable_shared_from_this<rtc_signaling>
+{
+public:
+	using offers_handler = std::function<void(error_code const&, std::vector<rtc_offer> const&)>;
+	using rtc_stream_handler = std::function<void(peer_id const &pid, rtc_stream_init&)>;
+
+	explicit rtc_signaling(io_context& ioc, torrent* t, rtc_stream_handler handler);
+	~rtc_signaling();
+	rtc_signaling& operator=(rtc_signaling const&) = delete;
+	rtc_signaling(rtc_signaling const&) = delete;
+	rtc_signaling& operator=(rtc_signaling&&) noexcept = delete;
+	rtc_signaling(rtc_signaling&&) noexcept = delete;
+
+	alert_manager& alerts() const;
+
+	void close();
+	void generate_offers(int count, offers_handler handler);
+	void process_offer(rtc_offer const& offer);
+	void process_answer(rtc_answer const& answer);
+
+#ifndef TORRENT_DISABLE_LOGGING
+	bool should_log() const;
+	void debug_log(const char* fmt, ...) const noexcept TORRENT_FORMAT(2,3);
+#endif
+
+private:
+	using description_handler = std::function<void(error_code const&, std::string const& description)>;
+	using waitable_timer = boost::asio::basic_waitable_timer<std::chrono::steady_clock>;
+
+	struct connection
+	{
+		explicit connection(io_context& ioc) : timer(ioc) {}
+
+		std::shared_ptr<rtc::PeerConnection> peer_connection;
+		std::shared_ptr<rtc::DataChannel> data_channel;
+		std::optional<peer_id> pid;
+
+		waitable_timer timer;
+	};
+
+	rtc_offer_id generate_offer_id() const;
+
+	connection& create_connection(rtc_offer_id const& offer_id, description_handler handler);
+	void on_generated_offer(error_code const& ec, rtc_offer offer);
+	void on_generated_answer(error_code const& ec, rtc_answer answer, rtc_offer offer);
+	void on_data_channel(error_code const& ec, rtc_offer_id offer_id, std::shared_ptr<rtc::DataChannel> dc);
+
+	io_context& m_io_context;
+	torrent* m_torrent;
+	const rtc_stream_handler m_rtc_stream_handler;
+
+	std::unordered_map<rtc_offer_id, connection, rtc_offer_id_hash> m_connections;
```

Comment:
I'm a bit surprised you need to specify a custom hash function here. Are you sure you do?
If you do, I would expect you could just say: `std::hash<rtc_offer_id>` rather than using your own class.

Reply:
The reason is there is no `std::hash` for `std::vector`, however there is one defined in boost. It's true that the struct is definitely overkill though, I'll replace it by `boost::hash<rtc_offer_id>`.